File: 278. First Bad Version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

      public class Solution extends VersionControl {
        public int firstBadVersion(int n) {
            // binary search
            // O(log(n))
            int start = 1;
            int end = n;
            int mid = 1;
            while (start <= end) {
                mid = start / 2 + end / 2;
                // mid and mid+1 are both bad. turned bad before mid+1
                if (isBadVersion(mid) && isBadVersion(mid + 1)) {
                    end = mid;
                // mid and mid+1 are both good. turned bad after mid
                } else if (!isBadVersion(mid) && !isBadVersion(mid + 1)) {
                    start = mid + 1;
                } else {
                    return mid + 1;
                }
            }
            // will not be executed
            return -1;
        }
    }

File: 278. First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        # base case
        if isBadVersion(1):
            return 1
        
        # binary search to find the last good ver and the first bad ver
        L = 1
        R = n
        while L < R:
            print(L, R)
            if L + 1 == R:
                return R
            mid = (L + R) // 2
            if (isBadVersion(mid)):
                R = mid
            else:
                L = mid